components/batch-system/src/batch.rs

```diff
@@ -571,9 +635,10 @@ where
         info!("shutdown batch system {}", name_prefix);
         self.router.broadcast_shutdown();
         let mut last_error = None;
-        for h in self.workers.drain(..) {
-            debug!("waiting for {}", h.thread().name().unwrap());
-            if let Err(e) = h.join() {
+        for (_, mut v) in self.workers.lock().unwrap().drain() {
+            let handle = v.0.take().unwrap();
+            debug!("waiting for {}", handle.thread().name().unwrap());
+            if let Err(e) = handle.join() {
```

Comment:
This may introduce deadlock, if `handle.join()` run before `Poller::drop`, `Poller::drop` will wait for the lock while this thread holding the lock and waiting for `Poller::drop`.

Reply:
Fixed, thanks!